src/flint/types/fmpz_mod_poly.pyx

```diff
@@ -0,0 +1,1529 @@
+
+        if typecheck(other, fmpz_mod_poly):
+            res = (self.ctx == (<fmpz_mod_poly>other).ctx) and \
+                  fmpz_mod_poly_equal(self.val, (<fmpz_mod_poly>other).val, self.ctx.mod.val)
+            if op == 2:
+                return res
+            else:
+                return not res
+        else:
+            return NotImplemented
+
+    def __getitem__(self, long i):
+        cdef fmpz_mod x
+        x = fmpz_mod.__new__(fmpz_mod)
+        x.ctx = self.ctx.mod
+        if i < 0:
+            return x
+        fmpz_mod_poly_get_coeff_fmpz(
+            x.val, self.val, i, self.ctx.mod.val
+        )
+        return x
+
+    def __setitem__(self, long i, x):
+        if i < 0:
+            raise ValueError("cannot assign to index < 0 of polynomial")
+        v = self.ctx.mod.any_as_fmpz_mod(x)
+        if v is NotImplemented:
+            raise TypeError
+        fmpz_mod_poly_set_coeff_fmpz(
+            self.val, i, (<fmpz_mod>v).val, self.ctx.mod.val
+        )
+
+    def __len__(self):
+        return fmpz_mod_poly_length(self.val, self.ctx.mod.val)
+
+    def __hash__(self):
+        return hash(tuple(self.coeffs()))
+
+    def __call__(self, input):
+        cdef fmpz_mod res
+
+        val = self.ctx.mod.any_as_fmpz_mod(input)
+        if val is NotImplemented:
+            raise TypeError(f"Cannot evaluate the polynomial with input: {input}")
+
+        res = fmpz_mod.__new__(fmpz_mod)
+        res.ctx = self.ctx.mod
+        fmpz_mod_poly_evaluate_fmpz(res.val, self.val, (<fmpz_mod>val).val, self.ctx.mod.val)
+        return res
+    
+    def multipoint_evaluation(self, vals):
+        """
+        Returns a list of values computed from evaluating
+        ``self`` at the ``n`` values given in the vector ``val``
+        
+        TODO: We could allow passing as an optional input the
+        subproduct tree, which would allow for faster, repeated
+        multipoint evaluations
+
+        TODO!! This currently segfaults for very large input,
+        suggesting a memory leak. Is this something our side of
+        on the FLINT side?
+
+            >>> R = fmpz_mod_poly_ctx(163)
+            >>> f = R([1,2,3,4,5])
+            >>> [f(x) for x in [-1,-2,-3]]
+            [fmpz_mod(3, 163), fmpz_mod(57, 163), fmpz_mod(156, 163)]
+            >>> f.multipoint_evaluation([-1,-2,-3])
+            [fmpz_mod(3, 163), fmpz_mod(57, 163), fmpz_mod(156, 163)]
+        """
+        cdef fmpz_mod f
+
+        if not isinstance(vals, (list, tuple)):
+            raise ValueError("Input must be a list of points")
+        
+        n = len(vals)
+        xs = <fmpz_struct *> libc.stdlib.malloc(cython.sizeof(fmpz_struct) * n)
+        for i in range(n):
+            val = self.ctx.mod.any_as_fmpz_mod(vals[i])
+            if val is NotImplemented:
+                libc.stdlib.free(xs)
+                raise ValueError(f"Unable to cast {val} to an `fmpz_mod`")
+            xs[i] = (<fmpz_mod>(val)).val[0]
```

Comment:
There are several problems here.

Firstly calling `free` is not enough to deallocate the fmpz because `fmpz_clear` needs to be used. However for that to work the array `xs` needs to own the fmpz which is the second problem...

Secondly assigning an fmpz struct like this is not safe. For large values the fmpz struct is actually just a pointer (or at least is treated like a pointer by Flint functions). Here you are copying that pointer into an array. What is pointed to though is owned by `val` which is a Python object. On each iteration of the loop `val` is reassigned which means that the previous Python object referenced  by `val` gets decref'ed and will then be deallocated. The `fmpz_mod.__dealloc__` method will call `fmpz_clear` which will free the object that was pointed to. At that point the pointer that is effectively stored in `xs[i]` is pointing to freed memory.

A simple fix is to use `fmpz_init_set` here rather than just assigning like `xs[i] = `. That will copy the `fmpz_t` including allocating a new copy of what is pointed to that would then be "owned" by the `xs` array. Then when `val` is deallocated `xs[i]` will still point to valid owned memory. We need to make sure then though that anywhere we do `free(xs)` we first have a loop that goes through calling `fmpz_clear`.

Ideally though we would bypass creating the Python object `val` in the first place if the target is to make an array of `fmpz_t` at the C level. What would be needed is something a bit like `fmpz_set_any_ref` that would initialise an `fmpz_t` by doing the same sort of conversions that `any_as_fmpz_mod` is doing.

I think that Cython's memoryviews or rather Cython arrays could simplify deallocation here:
http://docs.cython.org/en/latest/src/userguide/memoryviews.html#cython-arrays

I think that a Cython array would be automatically deallocated by Cython on exit from this function but a callback would be needed that could call `fmpz_clear` for any already initialised values.

Reply:
I don't know if it is likely that we would want to have many functions using arrays of fmpz like this but if so then it might be worth creating a `cdef class fmpz_vec` type to represent that. Basically we just want something that can handle memory management for an array of fmpz but that can also expose a raw pointer to the array for passing into C functions.

Reply:
There are a few places where polynomials want lists of points and for FLINT this ultimately ends up wanting to be `* fmpz_struct` as input, I'm happy to make the `fmpz_vec` and `fmpz_mod_vec`, as needed for certain use-cases.

Honestly I got a big confused by all of this and so tried copying how `acb_poly` handles the allocation etc, but fundementally I suppose these types are different.

When you mention `fmpz_vec` would this be similar to the FLINT `*_vec` types? Or something different?

Reply:
> When you mention `fmpz_vec` would this be similar to the FLINT `*_vec` types?

Yes, something similar. We basically need it for the same reason that Flint does.

I would envisage something like:
```cython
cdef class fmpz_mod_vec:
    # I think that Cython will initialise these to 0, NULL and None automatically:
    cdef slong size
    cdef fmpz *vec
    cdef fmpz_mod_ctx ctx

    def __init__(self, fmpz_mod_ctx ctx, values):
        cdef slong size
        size = len(values)
        vec = _fmpz_vec_init(size)

        self.size = size
        self.vec = vec
        self.ctx = ctx

        for i in range(size):
            success = ctx.set_any_as_new_fmpz_mod(self.vec[i], values[i])
            if not success:
                raise TypeError(f"Cannot convert {type(values[i])} to fmpz_mod")

    def __dealloc__(self):
        if self.vec != NULL:
            _fmpz_vec_clear(self.vec, self.size)

    @staticmethod
    cdef new(fmpz_mod_ctx ctx, slong size):
        cdef fmpz_mod_vec obj
        cdef fmpz *vec
        obj = fmpz_mod_vec.__new__(fmpz_mod_vec)
        vec = _fmpz_vec_init(size) # set entries to zero

        obj.size = size
        obj.vec = vec
        obj.ctx = ctx

        return obj
            
```

Reply:
OK, so I haven't made a class (yet) but I have used `_fmpz_vec_init` and `_fmpz_vec_clear` which seems to have fixed the incorrect values and the segfaults, so I'm going to push this at some point and making the full class can be done in a later PR so that this one doesnt get too massive